src/system_database.ts

```diff
@@ -553,6 +554,69 @@ export class PostgresSystemDatabase implements SystemDatabase {
     }
   }
 
+  async forkWorkflow(originalWorkflowID: string, forkedWorkflowId: string, startStep: number = 1): Promise<string> {
+    const workflowStatus = await this.getWorkflowStatusInternal(originalWorkflowID);
+    if (workflowStatus === null) {
+      throw new DBOSNonExistentWorkflowError(`Workflow ${originalWorkflowID} does not exist`);
+    }
+
+    const inputs = await this.getWorkflowInputs(originalWorkflowID);
+    if (inputs === null) {
+      throw new DBOSNonExistentWorkflowError(`Workflow ${originalWorkflowID} does not exist`);
+    }
+
+    const workflowStatusInternal: WorkflowStatusInternal = {
+      ...workflowStatus,
+      workflowUUID: forkedWorkflowId,
+      status: StatusString.ENQUEUED,
+      workflowName: workflowStatus.workflowName,
+      workflowClassName: workflowStatus.workflowClassName,
+      workflowConfigName: workflowStatus.workflowConfigName,
+      queueName: INTERNAL_QUEUE_NAME,
+      authenticatedUser: workflowStatus.authenticatedUser,
+      authenticatedRoles: workflowStatus.authenticatedRoles,
+      request: workflowStatus.request,
+      executorId: globalParams.executorID,
+      applicationVersion: workflowStatus.applicationVersion,
+      applicationID: workflowStatus.applicationID,
+      createdAt: Date.now(),
+    };
+
+    await this.initWorkflowStatus(workflowStatusInternal, inputs);
+
+    if (startStep > 1) {
+      const query = `
+        INSERT INTO ${DBOSExecutor.systemDBSchemaName}.operation_outputs (
+        workflow_uuid,
+        function_id,
+        output,
+        error,
+        function_name,
+        child_workflow_id
+      )
+      SELECT
+        $1 AS workflow_uuid,
+        function_id,
+        output,
+        error,
+        function_name,
+        child_workflow_id
+        FROM ${DBOSExecutor.systemDBSchemaName}.operation_outputs
+        WHERE workflow_uuid = $2 AND function_id < $3
+      `;
+
+      await this.pool.query(query, [forkedWorkflowId, originalWorkflowID, startStep]);
+    }
+
```

Comment:
Let's move this part of the code to a separate PR, because this is incomplete and untested.